### ISSUE TITLE
Using a function as a default for a Callable in a child class fails

### ISSUE THREAD

**Issue by jvkersch** [enthought/traits#411]
I have a `HasTraits` class with some `Callable` traits, and a second class that derives from it. I was trying to set the defaults for these traits as class variables, but I noticed that sometimes these defaults didn't get applied. The following example will make this clear. Here I have two Callables in the base class, and one is initialized in the derived class with a builtin function and one with a regular function. The latter does not get applied.

```python
from traits.api import HasStrictTraits, Callable

def regular_fun():
    pass

builtin_fun = print

class A(HasStrictTraits):
    f1 = Callable()
    f2 = Callable()

class B(A):
    f1 = builtin_fun
    f2 = regular_fun

if __name__ == '__main__':
    b = B()
    assert b.f1 is not None
    assert b.f2 is not None
```

Under the tested version (4.6.0) the first assert passes (as expected), but the second assert fails (`b.f2` is None). 

This is because in `MetaHasTraitsObject` we check whether items in the class_dict are functions: https://github.com/enthought/traits/blob/b5cf152c446bb87d575b5ecdc4cda7f8b646d96f/traits/has_traits.py#L550-L553

Builtin functions don't pass this check, but regular functions do. So `f2` gets treated in this block and ultimately ignored, but `f1` skips this block and ends up in the `else` where we look at inherited class traits:

https://github.com/enthought/traits/blob/b5cf152c446bb87d575b5ecdc4cda7f8b646d96f/traits/has_traits.py#L577-L590

Here, `builtin_fun` gets correctly set as the default for `f1`, but `f2` has already been treated and therefore keeps it None default.

**Comment by mdickinson** [comment]
So given that we're calling this type `Callable`, the obvious check would be to use Python's `callable()` built-in function. I can't think of any obvious reason not to do this. (Historical note: `callable` was removed in Python 3.0, but re-instated in Python 3.2.)

**Comment by mdickinson** [comment]
Sorry, that was a dumb comment based on an overly-quick reading of the report; I think I understand better now. The issue has nothing to do with how we validate `Callable` objects - we're simply catching the user-defined function too early, on the basis that it *might* be an `on_trait_change`-decorated method, and then when we discover it isn't, not bothering to pass it down the line. So it looks as though the fix is simply to replace the check
```
isinstance( value, FunctionType ) or is_cython_func_or_method(value)
```
with 
```
(isinstance( value, FunctionType ) or is_cython_func_or_method(value) ) and hasattr(value, "on_trait_change")
```
(or some refactored version of the above). @jvkersch Does that sound right?

**Comment by mdickinson** [comment]
... though I'm a bit worried that once that fix is made, your code *still* won't work as expected because of Python interpreting `b.f2` as a class method and implicitly passing `self` as the first argument ...

**Comment by corranwebster** [comment]
I think that's more-or-less right (although you might be able to replace the first clause with a `callable` check).

It will have some odd consequences, as then
```
def B(A):
    def f1():
        pass
```
will (I think) have the method-ish looking function `f1` become the default value for the `f1` trait, rather than a proper method (in particular, I don't think it will be given a `self` argument when called.  I'd really like this to raise an exception, but at the core we can't tell the difference between `def f1` and `f1 = ...`

**Comment by corranwebster** [comment]
@mdickinson I don't think it would get treated as a method as the default value gets popped from the class dict:
https://github.com/enthought/traits/blob/b5cf152c446bb87d575b5ecdc4cda7f8b646d96f/traits/has_traits.py#L595

**Comment by mdickinson** [comment]
Yep, looks like we're good. With the suggested change:
```
Python 2.7.13 |Enthought, Inc. (x86_64)| (default, Mar  2 2017, 08:20:50) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from traits.api import *
>>> def square(x): return x*x
... 
>>> class A(HasStrictTraits):
...     foo = Callable()
... 
>>> class B(A):
...     foo = square
... 
>>> b = B()
>>> b.foo(3)
9
```

**Comment by jvkersch** [comment]
I agree that the suggested change is the way to go.

**Comment by mdickinson** [comment]
@jvkersch Do you have the bandwidth to make a PR?

**Comment by jvkersch** [comment]
I'll try, but it might have to wait until the weekend.

**Comment by rkern** [comment]
For what it's worth, even with this change committed, I'd still recommend people just redefine the trait in the subclass and write `foo = Callable(square)`. It immediately resolves in the reader's brain the question of whether or not Traits is going to do what is intended or leave it to Python semantics or just let it drop off the face of the Earth.

**Comment by mdickinson** [comment]
#412 is in the process of being reverted (in #413); re-opening this issue.

Related: enthought/codetools#22
